stdlib/LinearAlgebra/src/triangular.jl

```diff
@@ -752,6 +752,80 @@ for op in (:+, :-)
     end
 end
 
+function kron(A::UpperTriangular{T}, B::UpperTriangular{S}) where {T,S}
+    C = UpperTriangular(Matrix{promote_op(*, T, S)}(undef, _kronsize(A, B)))
+    return kron!(C, A, B)
+end
+
+function kron(A::LowerTriangular{T}, B::LowerTriangular{S}) where {T,S}
+    C = LowerTriangular(Matrix{promote_op(*, T, S)}(undef, _kronsize(A, B)))
+    return kron!(C, A, B)
+end
+
+function kron!(C::UpperTriangular, A::UpperTriangular, B::UpperTriangular)
+    size(C) == _kronsize(A, B) || throw(DimensionMismatch("kron!"))
+    _triukron!(C.data, A.data, B.data)
+    return C
+end
+
+function kron!(C::LowerTriangular, A::LowerTriangular, B::LowerTriangular)
+    size(C) == _kronsize(A, B) || throw(DimensionMismatch("kron!"))
+    _trilkron!(C.data, A.data, B.data)
+    return C
+end
+
+function _triukron!(C, A, B)
+    n_A = size(A, 1)
+    n_B = size(B, 1)
+    @inbounds for j = 1:n_A
+        jnB = (j - 1) * n_B
+        for i = 1:(j-1)
+            Aij = A[i, j]
+            inB = (i - 1) * n_B
+            for l = 1:n_B
+                for k = 1:l
+                    C[inB+k, jnB+l] = Aij * B[k, l]
+                end
+                for k = 1:(l-1)
+                    C[inB+l, jnB+k] = zero(eltype(C))
+                end
+            end
+        end
+        Ajj = A[j, j]
+        for l = 1:n_B
+            for k = 1:l
+                C[jnB+k, jnB+l] = Ajj * B[k, l]
+            end
+        end
+    end
+end
+
+function _trilkron!(C, A, B)
+    n_A = size(A, 1)
+    n_B = size(B, 1)
+    @inbounds for j = 1:n_A
+        jnB = (j - 1) * n_B
+        Ajj = A[j, j]
+        for l = 1:n_B
+            for k = l:n_B
+                C[jnB+k, jnB+l] = Ajj * B[k, l]
+            end
+        end
+        for i = (j+1):n_A
+            Aij = A[i, j]
+            inB = (i - 1) * n_B
+            for l = 1:n_B
+                for k = l:n_B
+                    C[inB+k, jnB+l] = Aij * B[k, l]
+                end
+                for k = (l+1):n_B
+                    C[inB+l, jnB+k] = zero(eltype(C))
```

Comment:
```suggestion
                    C[inB+l, jnB+k] = _zero(C, inB+l, jnB+k)
```
and similarly, elsewhere. Currently, on this PR, the following is broken:
```julia
julia> U2 = UpperTriangular(reshape(fill(ones(2,2),4), 2, 2))
2×2 UpperTriangular{Matrix{Float64}, Matrix{Matrix{Float64}}}:
 [1.0 1.0; 1.0 1.0]  [1.0 1.0; 1.0 1.0]
         ⋅           [1.0 1.0; 1.0 1.0]

julia> kron(U2, U2)
ERROR: MethodError: no method matching zero(::Type{Matrix{Float64}})
The function `zero` exists, but no method is defined for this combination of argument types.

Closest candidates are:
  zero(::Type{Union{}}, Any...)
   @ Base number.jl:310
  zero(::Type{Missing})
   @ Base missing.jl:104
  zero(::Type{Dates.DateTime})
   @ Dates ~/Downloads/julia-674fcf7a91-linux-x86_64/julia-674fcf7a91/share/julia/stdlib/v1.11/Dates/src/types.jl:458
  ...

Stacktrace:
 [1] _triukron!(C::Matrix{Matrix{Float64}}, A::Matrix{Matrix{Float64}}, B::Matrix{Matrix{Float64}})
   @ LinearAlgebra ~/Downloads/julia-674fcf7a91-linux-x86_64/julia-674fcf7a91/share/julia/stdlib/v1.11/LinearAlgebra/src/triangular.jl:790
 [2] kron!
   @ ~/Downloads/julia-674fcf7a91-linux-x86_64/julia-674fcf7a91/share/julia/stdlib/v1.11/LinearAlgebra/src/triangular.jl:767 [inlined]
 [3] kron(A::UpperTriangular{Matrix{Float64}, Matrix{Matrix{Float64}}}, B::UpperTriangular{Matrix{Float64}, Matrix{Matrix{Float64}}})
   @ LinearAlgebra ~/Downloads/julia-674fcf7a91-linux-x86_64/julia-674fcf7a91/share/julia/stdlib/v1.11/LinearAlgebra/src/triangular.jl:757
 [4] top-level scope
   @ REPL[51]:1
```
I think this should be fixed by this change.

Reply:
Ops, indeed, thanks for catching that. However, I don't think this is a full solution; if we do a `Matrix{Matrix{Matrix}}`, that is,
```julia
c=reshape(fill(ones(2,2),4), 2, 2)
d = [c for i=1:2, j=1:2]
```
then we still have a crash, and currently `kron` can handle this.

Reply:
Upon further investigation, I saw that there is also a problem with `Symmetric`:
```julia
julia> s=Symmetric(reshape(fill(ones(2,2),4), 2, 2))
2×2 Symmetric{AbstractMatrix, Matrix{Matrix{Float64}}}:
 [1.0 1.0; 1.0 1.0]  [1.0 1.0; 1.0 1.0]
 [1.0 1.0; 1.0 1.0]  [1.0 1.0; 1.0 1.0]

julia> kron(s,s)
ERROR: MethodError: no method matching symmetric_type(::Type{Any})
The function `symmetric_type` exists, but no method is defined for this combination of argument types.

Closest candidates are:
  symmetric_type(::Type{T}) where T<:Number
   @ LinearAlgebra ~/local/julia/usr/share/julia/stdlib/v1.11/LinearAlgebra/src/symmetric.jl:95
  symmetric_type(::Type{T}) where {S<:(AbstractMatrix), T<:AbstractMatrix{S}}
   @ LinearAlgebra ~/local/julia/usr/share/julia/stdlib/v1.11/LinearAlgebra/src/symmetric.jl:92
  symmetric_type(::Type{T}) where {S<:Number, T<:AbstractMatrix{S}}
   @ LinearAlgebra ~/local/julia/usr/share/julia/stdlib/v1.11/LinearAlgebra/src/symmetric.jl:89
  ...

Stacktrace:
 [1] symmetric_type(::Type{Matrix{Any}})
   @ LinearAlgebra ~/local/julia/usr/share/julia/stdlib/v1.11/LinearAlgebra/src/symmetric.jl:87
 [2] Symmetric(A::Matrix{Any}, uplo::Symbol)
   @ LinearAlgebra ~/local/julia/usr/share/julia/stdlib/v1.11/LinearAlgebra/src/symmetric.jl:62
 [3] kron(A::Symmetric{AbstractMatrix, Matrix{Matrix{Float64}}}, B::Symmetric{AbstractMatrix, Matrix{Matrix{Float64}}})
   @ LinearAlgebra ~/local/julia/usr/share/julia/stdlib/v1.11/LinearAlgebra/src/symmetric.jl:532
 [4] top-level scope
   @ REPL[105]:1

```
Note that the `eltype(s)` is `AbstractMatrix`. When this gets fed into `promote_op` it goes crazy.

Honestly I think it's better to just change the methods to
```julia
function kron(A::Symmetric{T}, B::Symmetric{S}) where {T<:Number,S<:Number}
```
and so on, so that we remove these recursive matrices from consideration.